README.md
# Overview of genebean/packer-templates

This repo contains my [Packer][packer] templates for building bases boxes used
in [Vagrant][vagrant] and [Docker][docker]. By default it will build the Vagrant
boxes in Virtualbox and Docker format. There are also definitons for VMware boxes
but these are untested due to issues getting Packer and VMware Player to play
nice with each other. The files in the repo are laid out
so that each OS and version combo has a top-level folder. Under each of these
folders are all the files needed to build that combo in several different
configurations including:
- `base`: the common base that the boxes below are built on top of
- `nocm`: a setup without any configuration management installed
- `puppet`: Puppet 3.x is installed
- `puppet-agent`: Puppet 4.x is installed (uses the new all-in-one pacakage)
- `rvm-193`: RVM and version 1.9.3 of Ruby are installed
- `rvm-221`: RVM and version 2.2.1 of Ruby are installed

The base version is a little different than the others. When used for a
virtual machine it does not produce a Vagrant box but rather creates a
VM that is used by all the other builds. The Docker version does produce
an image that is essentially identical to the "nocm" version. This is
done primarily for consistancy as the "nocm" version of each is what
will be shared publically.

Many of the ideas and concepts for these templates have been pulled
from the following two sources:
- [shiguredo/packer-templates][shiguredo/packer-templates] -
  I originally forked this repo but have since reworked it completely based on
  some tricks learned from my next source.
- [puppetlabs/puppetlabs-packer][puppetlabs/puppetlabs-packer] -
  I now utilize an idea gathered from here to build a base VM and then
  separately build each box, adding in needed modifications as needed as part of
  the build process.


# Usage

## Installing Packer

If you're using Homebrew:

```bash
$ brew tap homebrew/binary
$ brew install packer
```

Otherwise, download the latest packer from http://www.packer.io/downloads.html
and unzip the appropriate directory.


## Running Packer

```bash
$ git clone https://github.com/genebean/packer-templates.git
$ cd packer-templates
```

### Build all boxes in all folders

```bash
$ ./run-all-builds.sh
```

By default, this will build all boxes by looping over the top
level folders and calling `build-all.sh` in each. If you want
to speed things up a bit you can run all the `build-all.sh`
scripts in parallel like this:

```bash
$ ./run-all-builds.sh -p
```


### Build all boxes in a single folder

```bash
$ cd some-os-version-combo
$ ./build-all.sh
```

### Build one box at a time

```bash
$ packer build template-base.json # must be run before other templates
# Replace template-some-type with the type of box you want to build.
$ packer build template-some-type.json
```

### Build for single platform

```bash
# replace "centos-7" with the name of the folder you are in
# replace "puppet" with the type of box you want

# for Virtualbox:
$ packer build -only=virtualbox-base-centos-7 template-base.json
$ packer build -only=virtualbox-vagrant-puppet-centos-7 template-puppet.json

# for Docker:
$ packer build -only=docker-base-centos-7 template-base.json
$ packer build -only=docker-puppet-centos-7 template-puppet.json

# for VMware:
$ packer build -only=vmware-base-centos-7 template-base.json
$ packer build -only=vmware-vagrant-puppet-centos-7 template-puppet.json
```


### Supported Packer version

This templates was tested using a packer 0.8.6


# Development

If you would like to make new templates or change these to have your name
instead of `genebean` then you will want to take note not of the variables
in the templates and build scripts. Below are some details to get you
started.

## Build Scripts

This script has just two things you may want to change:

```bash
# this is in build-all.sh

# this is the first part of each box's name
box_prefix='centos-7'
docker_user='genebean'
```

If you are making a new OS and version combo then you will  want to update
the `box_prefix` value. If you want the Docker images to have your name
associated with then then you will want to update `docker_user`.

## Templates

All the templates utilize variables for most things that you might want
to change. These variables are located at the top of the file. Details
specific to each template are below. Remember that any of these can be
updated via the command line also as an argument to `packer`.

### template-base

This template sets things up for all the subsequent builds and has the
following settings:

```bash
"variables": {
  "os": "centos-7",                   # This should match the folder you are in
  "docker_image": "centos:centos7",   # This is the Docker image to base your images on
  "docker_user": "genebean",          # This is the name that will be used when exporting the Docker image
  "build_name": "base",               # This is the name of the template you are in
  "iso_url": "http://[...].iso",      # The url or file path of the ISO to use
  "iso_checksum": "fssafsa"           # The sha1 checksum of the ISO above
}
```

### template-nocm

This template builds a box out of the base system without any further
modififications.

```bash
"variables": {
  "os": "centos-7",                          # This should match the folder you are in
  "docker_image": "genebean/centos-7-base",  # Builds this image based on the base one
  "docker_user": "genebean",                 # This is the name that will be used when exporting the Docker image
  "build_name": "nocm"                       # This is the name of the template you are in
}
```

### template-puppet / template-rvm-193 / etc...

These templates build a box out of the base system. They build a box
by running a script during provisioning that has the same name as
the template (`build_name`). All the specialized boxes are built
this way. The only difference in any of these templates is the build
name.

```bash
"variables": {
  "os": "centos-7",                          # This should match the folder you are in
  "docker_image": "genebean/centos-7-base",  # Builds this image based on the base one
  "docker_user": "genebean",                 # This is the name that will be used when exporting the Docker image
  "build_name": "puppet"                     # This is the name of the template you are in
}
```


## What's next

The Ubuntu 14.04 folder currently is just a carry-over from the old setup. I
plan to build this one out like the CentOS ones.

Also, I plan to add in Windows templates and, possibly, support for KVM.


[docker]: https://www.docker.com
[shiguredo/packer-templates]: https://github.com/shiguredo/packer-templates
[packer]: https://packer.io
[puppetlabs/puppetlabs-packer]: https://github.com/puppetlabs/puppetlabs-packer
[vagrant]: https://www.vagrantup.com


CHANGELOG.md
[![GitHub tag][gh-tag-img]][gh-link]

## 2015-11-30 Release 1.0.0  
* Initial release revised repo

[gh-tag-img]: https://img.shields.io/github/tag/genebean/packer-templates.svg?label=newest%20tag
[gh-link]: https://github.com/genebean/packer-templates


centos-7/build-all.sh
#!/bin/bash
DIR="$(dirname "$(perl -e 'use Cwd "abs_path";print abs_path(shift)' "$0")")"
TMPDIR="$(perl -e 'use Cwd "abs_path";print abs_path(shift)' "../tmp")"
cd $DIR

rm -f boxes/*
rm -rf output-*

# this is the first part of each box's name
box_prefix='centos-7'
docker_user='genebean'

# build base VM that is used for all boxes
packer build -force -except=vmware-base-${box_prefix} template-base.json
echo 'testing Docker image...'
docker run --name ${box_prefix}-base-hello-world ${docker_user}/${box_prefix}-base /bin/cat /etc/motd || exit 1
docker rm ${box_prefix}-base-hello-world

# ensure the base box was built
if [ ! -f "output-virtualbox-base-${box_prefix}/packer-virtualbox-base-${box_prefix}.ovf" ]; then
  echo "Error: can't find output-virtualbox-base-${box_prefix}/packer-virtualbox-base-${box_prefix}.ovf"
  exit 1
fi

# build each box
for box in 'nocm' 'puppet' 'puppet-agent' 'rvm-193' 'rvm-221'; do
  packer build -force -except=vmware-vagrant-${box}-${box_prefix} template-${box}.json

  # check if the box was built
  if [ ! -f "boxes/${box_prefix}-${box}-virtualbox.box" ]; then
    # if the box wasn't built try one more time before failing
    echo "trying again to build ${box_prefix}-${box}-virtualbox.box"
    packer build -force -except=vmware-vagrant-${box}-${box_prefix} template-${box}.json
  fi

  # if the box still does not exist then fail.
  if [ ! -f "boxes/${box_prefix}-${box}-virtualbox.box" ]; then
    echo "building ${box_prefix}-${box}-virtualbox.box failed."
    exit 1
  fi

  echo "testing ${box_prefix}-${box}-virtualbox.box with Vagrant"
  sleep 2

  vagrant box add boxes/${box_prefix}-${box}-virtualbox.box --name ${box_prefix}-${box} -f  || exit 1
  vagrant init -m ${box_prefix}-${box}  || exit 1
  vagrant up || exit 1
  sleep 2
  vagrant destroy -f || exit 1
  vagrant box remove ${box_prefix}-${box} -f || exit 1

  echo 'removing files made by Vagrant...'
  rm -rf .vagrant
  rm -f Vagrantfile

  echo 'testing Docker image...'
  docker run --name ${box_prefix}-${box}-hello-world ${docker_user}/${box_prefix}-${box} /bin/echo 'Hello world' || exit 1
  docker rm ${box_prefix}-${box}-hello-world
done

tree
docker images

echo 'all boxes built.'
